source/Halibut/Transport/ConnectionPool.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Runtime.Serialization;
+using Halibut.Diagnostics;
+using Halibut.Logging;
+
+namespace Halibut.Transport
+{
+    public class ConnectionPool<TKey, TPooledResource>
+        where TPooledResource : class, IPooledResource
+    {
+        readonly ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>> pool = new ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>>();
+
+        public int GetTotalConnectionCount()
+        {
+            return pool.Values.Sum(v => v.Count);
+        }
+
+        public TPooledResource Take(TKey endPoint)
+        {
+            while (true)
+            {
+                var connections = pool.GetOrAdd(endPoint, i => new ConcurrentDictionary<TPooledResource, byte>());
+                AtomicTake(connections, out var connection);
+
+                if (connection == null || !connection.HasExpired())
+                    return connection;
+
+                connection.Dispose();
```

Comment:
This should probably have a try/catch around it so that a faulted connection doesn't fail this new connection attempt

Reply:
`ConnectionPool` is generic and only works with `IPooledResource`. I'm not sure what you mean by faulted connection or how to detect that from the interface.

Reply:
`IPooledResource` is really a `SecureConnection`. The Dispose method of that could throw an exception (although SocketException is already covered). All I meant was that a failure in this dispose call shouldn't cause the call to this method to fail.